pkg/engine/factories/contextloaderfactory.go

```diff
@@ -8,79 +8,131 @@ import (
-	return func(policy kyvernov1.PolicyInterface, rule kyvernov1.Rule) engineapi.ContextLoader {
-		return &contextLoader{
-			logger:     logging.WithName("DefaultContextLoaderFactory"),
-			cmResolver: cmResolver,
+type ContextLoaderFactoryOptions func(*contextLoader)
+
+func DefaultContextLoaderFactory(opts ...ContextLoaderFactoryOptions) engineapi.ContextLoaderFactory {
+	return func(_ kyvernov1.PolicyInterface, _ kyvernov1.Rule) engineapi.ContextLoader {
+		cl := &contextLoader{
+			logger:       logging.WithName("DefaultContextLoaderFactory"),
+			initializers: make([]engineapi.Initializer, 0),
+		}
+		for _, o := range opts {
+			o(cl)
+		}
+		if cl.jp == nil {
+			cl.jp = jmespath.NewWithDefaults()
 		}
+		return cl
+	}
+}
+
+// WithAPIClient enables loading of APICall context entries
+func WithAPIClient(client engineapi.RawClient) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.client = client
+	}
+}
+
+// WithRegistryClientFactory enables loading of ImageData context entries
+func WithRegistryClientFactory(rcf engineapi.RegistryClientFactory) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.rclientFactory = rcf
+	}
+}
+
+// WithConfigMapResolver enables loading of ConfigMap context entries
+func WithConfigMapResolver(cmResolver engineapi.ConfigmapResolver) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.cmResolver = cmResolver
+	}
+}
+
+func WithInitializer(initializer engineapi.Initializer) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.initializers = append(cl.initializers, initializer)
+	}
+}
+
+// WithJMESPath sets the JMESPath engine
+func WithJMESPath(jp jmespath.Interface) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.jp = jp
 	}
 }
 
 type contextLoader struct {
-	logger     logr.Logger
-	cmResolver engineapi.ConfigmapResolver
+	logger         logr.Logger
+	jp             jmespath.Interface
+	client         engineapi.RawClient
+	rclientFactory engineapi.RegistryClientFactory
+	cmResolver     engineapi.ConfigmapResolver
+	initializers   []engineapi.Initializer
 }
 
 func (l *contextLoader) Load(
 	ctx context.Context,
-	jp jmespath.Interface,
-	client engineapi.RawClient,
-	rclientFactory engineapi.RegistryClientFactory,
 	contextEntries []kyvernov1.ContextEntry,
 	jsonContext enginecontext.Interface,
 ) error {
+	for _, init := range l.initializers {
+		if err := init(jsonContext); err != nil {
+			return err
+		}
+	}
+
 	for _, entry := range contextEntries {
-		deferredLoader := l.newDeferredLoader(ctx, jp, client, rclientFactory, entry, jsonContext)
-		if deferredLoader == nil {
-			return fmt.Errorf("invalid context entry %s", entry.Name)
+		deferredLoader, err := l.newDeferredLoader(ctx, entry, jsonContext)
+		if err != nil {
+			return fmt.Errorf("failed to create deferred loader for context entry %s", entry.Name)
+		}
+
+		if deferredLoader != nil {
```

Comment:
it's strange that a loader is nil here, we should at least log something

Reply:
Since a class of loaders can be disabled, `nil` is a valid response and logged in the delegate method. See [this line as an example]( https://github.com/kyverno/kyverno/blob/e9513d0476c288fb00a838785753c515d320c14a/pkg/engine/factories/contextloaderfactory.go#L113).